resource_activity/models/resource_activity_registration.py

```diff
@@ -228,31 +232,33 @@ def action_refresh(self):
     def reserve_needed_resource(self):
         for registration in self:
             if registration.quantity_needed == 0:
+                # fixme needed because _compute_state depends on allocations
+                # and there is no allocations here
+                # fixme make it depend on quantity_needed
                 registration.state = "booked"
-            else:
-                free_resources = registration.resources_available.filtered(
-                    lambda record: record.state == "free"
-                )
-                for resource_available in free_resources:
-                    if (
-                        registration.quantity_needed - registration.quantity_allocated
-                        <= 0
-                    ):
-                        break
-                    resource_available.action_reserve()
-
-                (registration.resource_activity_id.registrations.action_refresh())
-        return True
+                return
+
+            # update available resources
+            registration.search_resources()
+            free_resources = registration.resources_available.filtered(
+                lambda record: record.state == "free"
+            )
+            free_resources[: registration.quantity_needed].action_reserve()
+
+            if registration.quantity_allocated >= registration.quantity_needed:
+                registration.state = registration.booking_type
```

Comment:
a comment would be nice, especially for the 2nd line, but also to explain how it is possible for `quantity_allocated` to be `>` than `quantity_needed`.